### ISSUE TITLE
softmax_aggr += res[1] in predict.py in fast mode slow due to half-precision variables on cpu

### ISSUE THREAD

**Issue by PeterKoopmans** [MIC-DKFZ/nnUNet#1238]
Dear Fabian

I'd like to work with nnunet in a somewhat realtime environment (as in: trying to get a segmentation in about 1 minute, 2 minutes is already starting to become a burden). I put some additional timing print statements in predict.py to try and spot the slower sections. 

When running without the all-in-gpu flag, I noticed that softmax aggregation was very fast, but obviously the predictions themselves (i.e. the convolution loop) was a lot slower. Running with all_in_gpu = True, the reverse was the case unfortunately however. The reason seems to be that softmax_aggr is initialised as a float16 array, and my cpu clearly does not support half-precision natively. 

Some timing results:
When softmax_aggr and res[1] both are float32 (i.e. non-gpu mode) it takes 0.24 secs to add them with the += statement

If they both are float16, it takes 3.6 seconds, (meaning the 5 folds start to take a serious toll on my total execution time)

if I set softmax_aggr to float32, and let python handle the conversion of res[1] it takes 1.15 secs
(if I recast res[1] myself with ".astype" it takes a fraction longer)

So my question:

Can I simply change the initialisation for softmax_aggr to speed up the process or will I likely break something down the line?

or, alternatively and possibly even better: as I'm working with a 3090 with more than double the VRAM that was used during training (on different machine)": can I tell/force the program to simply use 32bit precision on the GPU or will this slow things down immensely by not managing to leverage half-precision compute options? 

Thanks for any help you may be able to give, and for nnunet in general :-)

Best, Peter

**Comment by saikat-roy** [comment]
Hi @PeterKoopmans. Is this still an active issue for you?

**Comment by PeterKoopmans** [comment]
Hi Saikat,

I havent tested the new 2.0 release yet nor have I checked whether 1.0 has
had any updates adressing it. In the end I made some minor changes to my
own local nnunet copy where I simply initialise the aggregate variables to
be single precision as opposed to half. This made it really a lot faster.
Before the changes, it made the gpu-all option equally slow as the cpu
version: all the gains obtained inside the evaluation loop were undone by
the adding of the softmax results in CPU-unsuitable format, this no longer
is the case.

Cheers Peter
